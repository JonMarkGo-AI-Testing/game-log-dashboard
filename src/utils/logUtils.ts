import { LogEntry } from '../store/gameStore';

interface TimeSeriesData {
  timestamp: string;
  error: number;
  warning: number;
  info: number;
}

interface ErrorDistributionData {
  category: string;
  count: number;
}

export const prepareTimeSeriesData = (logs: LogEntry[]): TimeSeriesData[] => {
  const timeMap = new Map<string, { error: number; warning: number; info: number }>();
  
  if (logs.length === 0) {
    return [];
  }

  // Find the time range from the logs
  const timestamps = logs.map(log => new Date(log.timestamp).getTime());
  const latestTime = Math.max(...timestamps);
  const earliestTime = Math.max(latestTime - (24 / 2 * 60 * 60 * 1000), Math.min(...timestamps));
  
  // Initialize time slots based on the log data's time range - 10 minute increments
  for (let time = earliestTime; time <= latestTime; time += 600000) { // 10 minutes = 600000ms
    const date = new Date(time);
    const hours = String(date.getHours()).padStart(2, '0');
    const minutes = String(Math.floor(date.getMinutes() / 10) * 10).padStart(2, '0');
    const hourKey = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}-${String(date.getDate()).padStart(2, '0')}T${hours}:${minutes}`;
    timeMap.set(hourKey, { error: 0, warning: 0, info: 0 });
  }
  
  // Count logs in each time slot
  logs.forEach(log => {
    const date = new Date(log.timestamp);
    const hours = String(date.getHours()).padStart(2, '0');
    const minutes = String(Math.floor(date.getMinutes() / 10) * 10).padStart(2, '0');
    const hourKey = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}-${String(date.getDate()).padStart(2, '0')}T${hours}:${minutes}`;
    
    if (timeMap.has(hourKey)) {
      const current = timeMap.get(hourKey)!;
      current[log.type]++;
    }
  });
  
  return Array.from(timeMap.entries())
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([timestamp, counts]) => {
      const date = new Date(timestamp);
      return {
        timestamp: date.toLocaleTimeString('en-US', {
          hour: 'numeric',
          minute: '2-digit',
          hour12: true
        }),
        ...counts
      };
    });
};

export const prepareErrorDistribution = (logs: LogEntry[]): ErrorDistributionData[] => {
  const errorMap = new Map<string, number>();
  
  logs.filter(log => log.type === 'error').forEach(log => {
    const count = errorMap.get(log.category) || 0;
    errorMap.set(log.category, count + 1);
  });
  
  return Array.from(errorMap.entries())
    .map(([category, count]) => ({ category, count }))
    .sort((a, b) => b.count - a.count)
    .slice(0, 10); // Top 10 error categories
}; 
