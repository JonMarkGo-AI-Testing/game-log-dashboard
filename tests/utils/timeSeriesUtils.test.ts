import { prepareTimeSeriesData } from '../../src/utils/logUtils';
import { LogEntry } from '../../src/store/gameStore';

describe('prepareTimeSeriesData time window calculations', () => {
  // Helper to create mock logs with specific timestamps
  const createMockLog = (timestamp: number): LogEntry => ({
    id: Math.random().toString(),
    timestamp,
    type: 'error',
    severity: 'high',
    message: 'Test error',
    category: 'test'
  });

  it('should prevent incorrect time window calculations', () => {
    // Create a fixed time at noon UTC
    const now = new Date('2024-01-01T12:00:00Z').getTime();
    const logs: LogEntry[] = [
      createMockLog(now), // current time
      createMockLog(now - (25 * 60 * 60 * 1000)), // 25 hours ago (should be excluded)
      createMockLog(now - (23 * 60 * 60 * 1000)), // 23 hours ago (should be included)
      createMockLog(now - (12 * 60 * 60 * 1000)), // 12 hours ago (should be included)
    ];

    const result = prepareTimeSeriesData(logs);

    // Verify we have the expected number of 10-minute increments in 24 hours
    // 24 hours * 6 ten-minute increments per hour = 144 increments
    // Plus 1 for the inclusive end point
    expect(result.length).toBe(145);

    // Take a snapshot of the data structure
    expect({
      dataPointCount: result.length,
      // Sample points from throughout the day to verify coverage
      samplePoints: {
        start: result.slice(0, 3).map(r => r.timestamp),
        middle: result.slice(71, 74).map(r => r.timestamp),
        end: result.slice(-3).map(r => r.timestamp)
      },
      // Count the unique hours to verify 24-hour coverage
      uniqueHours: new Set(
        result.map(r => r.timestamp.match(/\d+:/)?.map(h => parseInt(h))[0])
      ).size
    }).toMatchSnapshot('24-hour-window-coverage');

    // Verify we have entries spanning close to 24 hours by checking hour markers
    const timeFormat = /(\d+):(\d+)\s(AM|PM)/;
    
    // Check first entry
    const firstEntry = result[0].timestamp.match(timeFormat);
    expect(firstEntry).toBeTruthy();
    
    // Check last entry
    const lastEntry = result[result.length - 1].timestamp.match(timeFormat);
    expect(lastEntry).toBeTruthy();

    if (firstEntry && lastEntry) {
      const getHour24 = (hour: string, meridiem: string) => {
        const h = parseInt(hour);
        if (meridiem === 'PM' && h !== 12) return h + 12;
        if (meridiem === 'AM' && h === 12) return 0;
        return h;
      };

      const firstHour = getHour24(firstEntry[1], firstEntry[3]);
      const lastHour = getHour24(lastEntry[1], lastEntry[3]);

      // If we're spanning 24 hours, the hours should be the same
      // (allowing for some timezone/DST variance)
      expect(firstHour).toBe(lastHour);
    }

    // Verify we have entries for all 24 hours
    const hours = new Set(
      result.map(r => {
        const match = r.timestamp.match(timeFormat);
        if (match) {
          return getHour24(match[1], match[3]);
        }
        return -1;
      })
    );
    
    // We should have entries for all 24 hours
    expect(hours.size).toBe(24);
  });
});

// Helper function to convert 12-hour format to 24-hour
function getHour24(hour: string, meridiem: string): number {
  const h = parseInt(hour);
  if (meridiem === 'PM' && h !== 12) return h + 12;
  if (meridiem === 'AM' && h === 12) return 0;
  return h;
}