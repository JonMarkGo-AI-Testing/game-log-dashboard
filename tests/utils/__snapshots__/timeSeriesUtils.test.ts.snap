// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepareTimeSeriesData time window calculations should prevent incorrect time window calculations: 24-hour-window-coverage 1`] = `
{
  "dataPointCount": 145,
  "samplePoints": {
    "end": [
      "6:40 AM",
      "6:50 AM",
      "7:00 AM",
    ],
    "middle": [
      "6:50 PM",
      "7:00 PM",
      "7:10 PM",
    ],
    "start": [
      "7:00 AM",
      "7:10 AM",
      "7:20 AM",
    ],
  },
  "uniqueHours": 12,
}
`;
